controller_manager/test/test_load_controller.cpp

```diff
@@ -183,192 +213,314 @@ TEST_F(TestLoadController, configure_controller)
+}
+
+TEST_F(TestLoadController, inactive_controller_cannot_be_cleaned_up)
+{
+  auto controller_if = cm_->load_controller(controller_name1, TEST_CONTROLLER_CLASS_NAME);
+  ASSERT_NE(controller_if, nullptr);
+  EXPECT_EQ(
+    cm_->configure_controller(controller_name1), controller_interface::return_type::OK);
+  // load and start controller with name1
+  {
+    auto switch_future = std::async(
+      std::launch::async,
+      &controller_manager::ControllerManager::switch_controller, cm_,
+      strvec{controller_name1}, strvec{},
+      STRICT, true, rclcpp::Duration(0, 0));
 
     ASSERT_EQ(
-      lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE,
-      abstract_test_controller1.c->get_current_state().id());
+      std::future_status::timeout,
+      switch_future.wait_for(std::chrono::milliseconds(100))) <<
+      "switch_controller should be blocking until next update cycle";
+    ControllerManagerRunner cm_runner(this);
+    EXPECT_EQ(
+      controller_interface::return_type::OK,
+      switch_future.get()
+    );
   }
+  // Stop controller
+  auto switch_future = std::async(
+    std::launch::async,
+    &controller_manager::ControllerManager::switch_controller, cm_,
+    strvec{}, strvec{controller_name1},
+    STRICT, true, rclcpp::Duration(0, 0));
+
+  ASSERT_EQ(
+    std::future_status::timeout,
+    switch_future.wait_for(std::chrono::milliseconds(100))) <<
+    "switch_controller should be blocking until next update cycle";
+  ControllerManagerRunner cm_runner(this);
+  EXPECT_EQ(
+    controller_interface::return_type::OK,
+    switch_future.get()
+  );
+
+  ASSERT_EQ(
+    lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE,
+    controller_if->get_current_state().id());
 
   std::shared_ptr<test_controller::TestController> test_controller =
-    std::dynamic_pointer_cast<test_controller::TestController>(abstract_test_controller1.c);
+    std::dynamic_pointer_cast<test_controller::TestController>(controller_if);
   size_t cleanup_calls = 0;
   test_controller->cleanup_calls = &cleanup_calls;
   // Configure from inactive state: controller can no be cleaned-up
   test_controller->simulate_cleanup_failure = true;
   EXPECT_EQ(cm_->configure_controller(controller_name1), controller_interface::return_type::ERROR);
   ASSERT_EQ(
     lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE,
-    abstract_test_controller1.c->get_current_state().id());
+    controller_if->get_current_state().id());
   EXPECT_EQ(0u, cleanup_calls);
-
-  // Configure from inactive state
-  test_controller->simulate_cleanup_failure = false;
-  EXPECT_EQ(
-    cm_->configure_controller(controller_name1), controller_interface::return_type::OK);
-  ASSERT_EQ(
-    lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE,
-    abstract_test_controller1.c->get_current_state().id());
-  EXPECT_EQ(1u, cleanup_calls);
 }
 
-TEST_F(TestLoadController, switch_controller_empty)
+TEST_F(TestLoadController, inactive_controller_cannot_be_configured)
```

Comment:
This test should be moved up before `can_not_configure_active_controller`, or the `can_not_configure_active_controller` should be at least after `can_stop_active_controller`.

In that was the tests would be like: “normal lifecycle”, "errors in lifecycle"